### PR TITLE
Autobuilder: fall back when os.Executable fails

### PR DESCRIPTION
**Pull request by smowton**
This can happen under tracing, perhaps because of https://github.com/github/codeql-tracer/issues/29

### CODE REVIEW

File: extractor/cli/go-autobuilder/go-autobuilder.go [hunk 1]
```diff
@@ -206,6 +207,51 @@ func checkVendor() bool {
...
+func getOsToolsSubdir() (string, error) {
+	switch runtime.GOOS {
+	case "darwin":
+		return "osx64", nil
+	case "linux":
+		return "linux64", nil
+	case "windows":
+		return "win64", nil
+	}
+	return "", errors.New("Unknown OS: " + runtime.GOOS)
```
Comment (sauyon) [inline extractor/cli/go-autobuilder/go-autobuilder.go:219]: Personally I like the `fmt.Errorf` more than `errors.New`, but that's entirely a style thing. Also, I think "unsupported" is more accurate here.

```suggestion
	return "", fmt.Errorf("Unsupported OS: %s", runtime.GOOS)
```

File: extractor/cli/go-autobuilder/go-autobuilder.go [hunk 2]
```diff
@@ -206,6 +207,51 @@ func checkVendor() bool {
...
+	mypath, err := os.Executable()
+	if err == nil {
+		return filepath.Dir(mypath), nil
+	}
+	log.Printf("Could not determine path of autobuilder: %v.\n", err)
+
+	// Fall back to rebuilding our own path from the extractor root:
+	extractorRoot := os.Getenv("CODEQL_EXTRACTOR_GO_ROOT")
+	if extractorRoot == "" {
+		return "", errors.New("CODEQL_EXTRACTOR_GO_ROOT not set")
```
Comment (sauyon) [inline extractor/cli/go-autobuilder/go-autobuilder.go:232]: We should probably have a similar message here as when `CODEQL_EXTRACTOR_GO_SOURCE_ARCHIVE_DIR` is unset below.

**Comment by smowton** [comment]
I'll improve the message in a followup since time it pressing

File: extractor/cli/go-autobuilder/go-autobuilder.go [hunk 3]
```diff
@@ -206,6 +207,51 @@ func checkVendor() bool {
 	return true
 }
 
+func getOsToolsSubdir() (string, error) {
```
[the remark this answers is not in the document]
Comment (smowton) [reply extractor/cli/go-autobuilder/go-autobuilder.go:210]: Yeah, but that was being filtered out, presumably also by the tracer. It was known to `autobuild.sh` but not `go-autobuild`.

[the remark this answers is not in the document]
Comment (aibaars) [reply extractor/cli/go-autobuilder/go-autobuilder.go:210]: @matt-gretton-dann The tracer shouldn't filter out `CODEQL_` variables, right?

[the remark this answers is not in the document]
Comment (matt-gretton-dann) [reply extractor/cli/go-autobuilder/go-autobuilder.go:210]: The tracer has a concept of critical variables which it tries to preserve in the environment.  But also when the tracer is turned off (as is the case when we trace go-autouild) we remove those critical variables.  I believe this was historically deemed appropriate to try and preserve the environment the user would expect

`CODEQL_PLATFORM` is one of those critical env. vars.

**Pull request merged**